src/devices/sound/hc55516.h

```diff
@@ -1,95 +1,212 @@
 	void digit_w(int digit);
 
-	/* sets the clock state (0 or 1, clocked on the rising edge) */
+	/* sets the clock state (0 or 1, clocked on the rising edge), common to all chips */
 	void clock_w(int state);
 
-	/* returns whether the clock is currently LO or HI */
-	int clock_state_r();
+	/* returns whether the clock is currently LO or HI, common to all chips */
+	virtual int clock_state_r();
 
 protected:
-	hc55516_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock);
+	cvsd_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock);
 
 	// device-level overrides
 	virtual void device_start() override;
 	virtual void device_reset() override;
+	//virtual void device_clock_changed() override;
 
 	// sound stream update overrides
 	virtual void sound_stream_update(sound_stream &stream, std::vector<read_stream_view> const &inputs, std::vector<write_stream_view> &outputs) override;
 
-	void start_common(uint8_t _shiftreg_mask, int _active_clock_hi);
+	// callbacks
+	devcb_write_line m_clock_state_push_cb; ///TODO: get rid of this, if you use it you should feel bad
+	devcb_read_line m_digin_pull_cb;
+	devcb_write_line m_digout_push_cb;
 
 	// internal state
-	sound_stream *m_channel;
-	int     m_active_clock_hi;
+	sound_stream *m_stream;
+	bool      m_active_clock_edge;
 	uint8_t   m_shiftreg_mask;
-
-	uint8_t   m_last_clock_state;
-	uint8_t   m_digit;
-	uint8_t   m_new_digit;
+	bool      m_last_clock_state;
+	bool      m_buffered_bit;
 	uint8_t   m_shiftreg;
-
 	stream_buffer::sample_t m_curr_sample;
 	stream_buffer::sample_t m_next_sample;
+	uint32_t  m_samples_generated;
 
-	uint32_t  m_update_count;
-
-	double  m_filter;
-	double  m_integrator;
+	// specific internal handler overrides, overridden by each chip
+	virtual void process_bit(bool bit, bool clock_state);
 
-	double  m_charge;
-	double  m_decay;
-	double  m_leak;
+	///TODO: get rid of these
+	inline bool is_external_oscillator();
+	inline bool is_clock_changed(bool clock_state);
+	inline bool is_active_clock_transition(bool clock_state);
+	inline bool current_clock_state();
 
-	inline int is_external_oscillator();
-	inline int is_active_clock_transition(int clock_state);
-	inline int current_clock_state();
-	void process_digit();
 };
 
 
-class mc3417_device : public hc55516_device
+class hc55516_device : public cvsd_device
 {
 public:
-	mc3417_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+	hc55516_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	// /FZ (partial reset) pull callback, ok to leave unconnected (we assume it is pulled high)
+	auto fzq_cb() { return m_fzq_pull_cb.bind(); }
+	// /FZ (partial reset) push
+	WRITE_LINE_MEMBER( fzq_w );
+	// AGC callback function, called to push the state if the AGC pin changes, ok to leave unconnected
+	auto agc_cb() { return m_agc_push_cb.bind(); }
```

Comment:
As with the other class, keep configuration member functions grouped, and keep them separate from “live” member functions.

Reply:
As with the other class, I've changed the order for the configuration members so the callbacks appear first and grouped together, as in 6525tpi.h